resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolverContext.java

```diff
@@ -214,13 +300,86 @@ void onResponse(final DnsQuestion question, AddressedEnvelope<DnsResponse, InetS
+        final DnsResponse res = envelope.content();
+
+        // Check if we have answers, if not this may be an non authority NS and so redirects must be handled.
+        if (res.count(DnsSection.ANSWER) == 0) {
+            AuthoritativeNameServerList serverNames = extractAuthoritativeNameServers(question.name(), res);
+
+            if (serverNames != null) {
+                List<InetSocketAddress> nameServers = new ArrayList<InetSocketAddress>(serverNames.size());
+                int additionalCount = res.count(DnsSection.ADDITIONAL);
+
+                for (int i = 0; i < additionalCount; i++) {
+                    final DnsRecord r = res.recordAt(DnsSection.ADDITIONAL, i);
+
+                    if ((r.type() == DnsRecordType.A && !parent.supportsARecords()) ||
+                            r.type() == DnsRecordType.AAAA && !parent.supportsAAAARecords()) {
+                        continue;
+                    }
+
+                    final String recordName = r.name();
+                    AuthoritativeNameServerList.AuthoritativeNameServer authoritativeNameServer =
+                            serverNames.remove(recordName);
+
+                    if (authoritativeNameServer == null) {
+                        // Not a server we are interested in.
+                        continue;
+                    }
+
+                    InetAddress resolved = parseAddress(r, recordName);
+                    if (resolved == null) {
+                        // Could not parse it, move to the next.
+                        continue;
+                    }
+
+                    nameServers.add(new InetSocketAddress(resolved, parent.dnsRedirectPort(resolved)));
+                    addNameServerToCache(authoritativeNameServer, resolved, r.timeToLive());
+                }
+
+                if (nameServers.isEmpty()) {
+                    promise.tryFailure(
+                            new UnknownHostException("Unable to find correct NS server for " + hostname));
+                } else {
+                    // Just use the same oder as we received the name servers for lookup.
+                    query(DnsServerAddresses.sequential(nameServers).stream(), question, promise);
+                }
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Returns the {@code {@link AuthoritativeNameServerList} which were included in {@link DnsSection#AUTHORITY}
+     * or {@code null} if non are found.
+     */
+    private static AuthoritativeNameServerList extractAuthoritativeNameServers(String questionName, DnsResponse res) {
+        int authorityCount = res.count(DnsSection.AUTHORITY);
+        if (authorityCount == 0) {
+            return null;
+        }
+
+        AuthoritativeNameServerList serverNames = new AuthoritativeNameServerList(questionName);
+        for (int i = 0; i < authorityCount; i++) {
+            serverNames.add(res.recordAt(DnsSection.AUTHORITY, i));
```

Comment:
How many items does the list usually have? I wonder what the performance would be against a plain null-terminated array or an array list.